# importing libraries

In [90]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [91]:
data = pd.read_excel('Data Science Case study Vindiata.xlsx',sheet_name=['AC characteristics','Operations','City pairs'])

In [92]:
AC_characteristics_df = data.get('AC characteristics')
Operations_df = data.get('Operations')
City_pairs_df = data.get('City pairs')

In [93]:
AC_characteristics_df.head()

Aircraft Costs     Unnamed: 1  Unnamed: 2         Unnamed: 3  \
0            USD            NaN         NaN                NaN   
1            NaN            NaN         NaN                NaN   
2            NaN  Aircraft Type  Range (Km)  Ave. Speed (km/h)   
3            NaN           A320        5000                800   
4            NaN           A330        8000                900   

        Unnamed: 4             Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  \
0              NaN                    NaN         NaN         NaN         NaN   
1              NaN                    NaN         NaN         NaN         NaN   
2  Number of Seats  Costs per flight hour         NaN         NaN         NaN   
3              150                   5000         NaN         NaN         NaN   
4              250                   7500         NaN         NaN         NaN   

   Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 14  Unnamed: 15  Unnamed: 16  Unnamed: 17  Unnamed: 18  
0          NaN          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN          NaN

In [94]:
AC_characteristics_df.drop('Aircraft Costs',axis=1,inplace=True)
AC_characteristics_df.drop(AC_characteristics_df.iloc[:,5:],axis=1,inplace=True)
AC_characteristics_df.drop(index=AC_characteristics_df.index[:2],inplace=True)
AC_characteristics_df.columns = AC_characteristics_df.iloc[0,:].values
AC_characteristics_df.drop([2],axis=0,inplace=True)
AC_characteristics_df.reset_index(drop=True,inplace=True)

In [95]:
AC_characteristics_df

Aircraft Type Range (Km) Ave. Speed (km/h) Number of Seats  \
0          A320       5000               800             150   
1          A330       8000               900             250   
2          B737       5000               800             150   
3          B747      10000               900             350   
4          Q400       1500               750              90   
5         ATR72       1000               650              75   

  Costs per flight hour  
0                  5000  
1                  7500  
2                  5100  
3                 12000  
4                  3500  
5                  2750

In [96]:
Operations_df

Airline A - hours flown     Unnamed: 1     Unnamed: 2  \
0                       NaN            NaN            NaN   
1                       NaN            NaN            NaN   
2                       NaN  Aircraft Name  Aircraft Type   
3                       NaN             A            A320   
4                       NaN              B           A330   
5                       NaN              C           B737   
6                       NaN              D           B747   
7                       NaN              E           Q400   
8                       NaN              F          ATR72   
9                       NaN              G           B747   
10                      NaN              H           Q400   
11                      NaN             I           ATR72   
12                      NaN              J           A330   
13                      NaN              K           B737   
14                      NaN              L           A320   
15                      NaN              M           A330   
16                      NaN              N           B737   
17                      NaN              O           A320   
18                      NaN              P           A320   
19                      NaN              Q           Q400   
20                      NaN              R          ATR72   
21                      NaN              S           A330   
22                      NaN              T           A330   
23                      NaN              U           B747   
24                      NaN              V           Q400   
25                      NaN              W          ATR72   
26                      NaN              X           A320   
27                      NaN              Y           A330   
28                      NaN              Z           B737   

             Unnamed: 3           Unnamed: 4           Unnamed: 5  \
0           Hours flown                  NaN                  NaN   
1                  2014                 2014                 2014   
2   2014-01-01 00:00:00  2014-02-01 00:00:00  2014-03-01 00:00:00   
3                   250                  251                  248   
4                   300                  301                  298   
5                   245                  246                  243   
6                   310                  311                  308   
7                   200                  201                  198   
8                   195                  196                  193   
9                   300                  301                  298   
10                  195                  196                  193   
11                  200                  201                  198   
12                  310                  311                  308   
13                  250                  251                  248   
14                  245                  246                  243   
15                  300                  301                  298   
16                  245                  246                  243   
17                  250                  251                  248   
18                  245                  246                  243   
19                  195                  196                  193   
20                  200                  201                  198   
21                  310                  311                  308   
22                  300                  301                  298   
23                  310                  311                  308   
24                  200                  201                  198   
25                  200                  201                  198   
26                  250                  251                  248   
27                  300                  301                  298   
28                  245                  246                  243   

             Unnamed: 6           Unnamed: 7           Unnamed: 8  \
0                   NaN          

In [97]:
Operations_df.drop('Airline A - hours flown',axis=1,inplace=True)
Operations_df.drop(Operations_df.iloc[:,14:],axis=1,inplace=True)
Operations_df.drop(index=Operations_df.index[:2],inplace=True)
Operations_df.columns = Operations_df.iloc[0,:].values
Operations_df.drop([2],axis=0,inplace=True)
Operations_df.reset_index(drop=True,inplace=True)

In [98]:
Operations_df

Aircraft Name Aircraft Type 2014-01-01 00:00:00 2014-02-01 00:00:00  \
0             A           A320                 250                 251   
1              B          A330                 300                 301   
2              C          B737                 245                 246   
3              D          B747                 310                 311   
4              E          Q400                 200                 201   
5              F         ATR72                 195                 196   
6              G          B747                 300                 301   
7              H          Q400                 195                 196   
8             I          ATR72                 200                 201   
9              J          A330                 310                 311   
10             K          B737                 250                 251   
11             L          A320                 245                 246   
12             M          A330                 300                 301   
13             N          B737                 245                 246   
14             O          A320                 250                 251   
15             P          A320                 245                 246   
16             Q          Q400                 195                 196   
17             R         ATR72                 200                 201   
18             S          A330                 310                 311   
19             T          A330                 300                 301   
20             U          B747                 310                 311   
21             V          Q400                 200                 201   
22             W         ATR72                 200                 201   
23             X          A320                 250                 251   
24             Y          A330                 300                 301   
25             Z          B737                 245                 246   

   2014-03-01 00:00:00 2014-04-01 00:00:00 2014-05-01 00:00:00  \
0                  248                 253                 243   
1                  298                 303                 293   
2                  243                 248                 238   
3                  308                 313                 303   
4                  198                 203                 193   
5                  193                 198                 188   
6                  298                 303                 293   
7                  193                 198                 188   
8                  198                 203                 193   
9                  308                 313                 303   
10                 248                 253                 243   
11                 243                 248                 238   
12                 298                 303                 293   
13                 243                 248                 238   
14                 248                 253                 243   
15                 243                 248                 238   
16                 193                 198                 188   
17                 198                 203                 193   
18                 308                 313                 303   
19                 298                 303                 293   
20                 308                 313                 303   
21                 198                 203                 193   
22                 198                 203                 193   
23                 248                 253                 243   
24                 298                 303                 293   
25                 243                 248                 238   

   2014-06-01 00:00:00 2014-07-01 00:00:00 2014-08-01 00:00:00  \
0                  245                 248                 246   
1                  295                 298                 296   
2                  240 

In [99]:
City_pairs_df

Airline B  - City Pairs considered   Unnamed: 1        Unnamed: 2  \
0                                 NaN          NaN               NaN   
1                                 NaN          NaN               NaN   
2                                 NaN  Origin City  Desitnation City   
3                                 NaN           AA                BB   
4                                 NaN           BB                CC   
5                                 NaN           CC                AA   
6                                 NaN           AA                DD   

                 Unnamed: 3     Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0                       NaN            NaN         NaN         NaN   
1                       NaN            NaN         NaN         NaN   
2  Pass. Demand \n(per day)  Distance (km)         NaN         NaN   
3                       420           3000         NaN         NaN   
4                       450           6500         NaN         NaN   
5                       300            400         NaN         NaN   
6                       300           1000         NaN         NaN   

   Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12  \
0         NaN         NaN         NaN          NaN          NaN          NaN   
1         NaN         NaN         NaN          NaN          NaN          NaN   
2         NaN         NaN         NaN          NaN          NaN          NaN   
3         NaN         NaN         NaN          NaN          NaN          NaN   
4         NaN         NaN         NaN          NaN          NaN          NaN   
5         NaN         NaN         NaN          NaN          NaN          NaN   
6         NaN         NaN         NaN          NaN          NaN          NaN   

   Unnamed: 13  Unnamed: 14  Unnamed: 15  Unnamed: 16  Unnamed: 17  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   
5          NaN          NaN          NaN          NaN          NaN   
6          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 18  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
5          NaN  
6          NaN

In [100]:
City_pairs_df.drop(City_pairs_df.columns[0],axis=1,inplace=True)
City_pairs_df.drop(City_pairs_df.iloc[:,4:],axis=1,inplace=True)
City_pairs_df.drop(index=City_pairs_df.index[:2],inplace=True)
City_pairs_df.columns = City_pairs_df.iloc[0,:].values
City_pairs_df.drop([2],axis=0,inplace=True)
City_pairs_df.reset_index(drop=True,inplace=True)

In [101]:
City_pairs_df

Origin City Desitnation City Pass. Demand \n(per day) Distance (km)
0          AA               BB                      420          3000
1          BB               CC                      450          6500
2          CC               AA                      300           400
3          AA               DD                      300          1000

In [102]:
dic = {}
total_hours=[]
for i in range(len(AC_characteristics_df)):
    df = Operations_df.groupby('Aircraft Type').sum()
    hours = df.iloc[i,1:].values.sum()
    cost = AC_characteristics_df[AC_characteristics_df['Aircraft Type'] == df.index[i]]['Costs per flight hour'].values[0]
    total = hours * cost
    total_hours.append(hours)
    dic[df.index[i]] = total


In [104]:
total_cost_df = pd.DataFrame()
total_cost_df['Aircraft Type'] = dic.keys()
total_cost_df['Total_hours'] = total_hours
total_cost_df['Total_Amount(in $)'] = dic.values()
total_cost_df

Aircraft Type  Total_hours  Total_Amount(in $)
0          A320        14875            74375000
1          A330        21834           163755000
2         ATR72         9536            26224000
3          B737        11816            60261600
4          B747        11037           132444000
5          Q400         9476            33166000

**These are the Total Amounts(in $) of each Aircraft Type for the year**

In [105]:
total_cost_df = total_cost_df.set_index('Aircraft Type')
total_cost_df = total_cost_df.reindex(index=AC_characteristics_df['Aircraft Type'])
total_cost_df = total_cost_df.reset_index()

In [106]:
total_cost_df

Aircraft Type  Total_hours  Total_Amount(in $)
0          A320        14875            74375000
1          A330        21834           163755000
2          B737        11816            60261600
3          B747        11037           132444000
4          Q400         9476            33166000
5         ATR72         9536            26224000

In [107]:
cost_per_seat_per_km_flown = (total_cost_df['Total_Amount(in $)'])/(total_cost_df['Total_hours'] * AC_characteristics_df['Ave. Speed (km/h)'] * AC_characteristics_df['Number of Seats'])

In [108]:
total_cost_df['cost_per_seat_per_km_flown(in $)'] = cost_per_seat_per_km_flown
total_cost_df

Aircraft Type  Total_hours  Total_Amount(in $)  \
0          A320        14875            74375000   
1          A330        21834           163755000   
2          B737        11816            60261600   
3          B747        11037           132444000   
4          Q400         9476            33166000   
5         ATR72         9536            26224000   

  cost_per_seat_per_km_flown(in $)  
0                         0.041667  
1                         0.033333  
2                           0.0425  
3                         0.038095  
4                         0.051852  
5                          0.05641

### From above dataframe we can see that A330 Aircraft Type is having the lowest cost per seat per km flown with cost of 0.033333 $

In [110]:
air_type=[]
no_trips=[]
cost_per_trip=[]
ori_des_pair=[]
for i in range(len(City_pairs_df)):
    d = AC_characteristics_df[AC_characteristics_df['Range (Km)'] >= City_pairs_df.loc[i,'Distance (km)']]
    for j in d.index:
        ori_des_pair.append(City_pairs_df.loc[i,'Origin City']+'-'+City_pairs_df.loc[i,'Desitnation City'])
        air_type.append(d.loc[j,'Aircraft Type'])
        no_trips.append(int(np.ceil(City_pairs_df.loc[i,'Pass. Demand \n(per day)'] / d.loc[j,'Number of Seats'])))
        cost_per_trip.append(d.loc[j,'Number of Seats'] * total_cost_df.loc[j,'cost_per_seat_per_km_flown(in $)'] * City_pairs_df.loc[i,'Distance (km)'])
        

In [111]:
final_df = pd.DataFrame()
final_df['ori_des_pair'] = ori_des_pair
final_df['air_type'] = air_type
final_df['NO. of trips will take'] = no_trips
final_df['cost_per_trip'] = cost_per_trip
final_df['Total_cost(in $)'] = final_df['NO. of trips will take'] * final_df['cost_per_trip']

In [112]:
final_df

ori_des_pair air_type  NO. of trips will take  cost_per_trip  \
0         AA-BB     A320                       3   18750.000000   
1         AA-BB     A330                       2   25000.000000   
2         AA-BB     B737                       3   19125.000000   
3         AA-BB     B747                       2   40000.000000   
4         BB-CC     A330                       2   54166.666667   
5         BB-CC     B747                       2   86666.666667   
6         CC-AA     A320                       2    2500.000000   
7         CC-AA     A330                       2    3333.333333   
8         CC-AA     B737                       2    2550.000000   
9         CC-AA     B747                       1    5333.333333   
10        CC-AA     Q400                       4    1866.666667   
11        CC-AA    ATR72                       4    1692.307692   
12        AA-DD     A320                       2    6250.000000   
13        AA-DD     A330                       2    8333.333333   
14        AA-DD     B737                       2    6375.000000   
15        AA-DD     B747                       1   13333.333333   
16        AA-DD     Q400                       4    4666.666667   
17        AA-DD    ATR72                       4    4230.769231   

    Total_cost(in $)  
0       56250.000000  
1       50000.000000  
2       57375.000000  
3       80000.000000  
4      108333.333333  
5      173333.333333  
6        5000.000000  
7        6666.666667  
8        5100.000000  
9        5333.333333  
10       7466.666667  
11       6769.230769  
12      12500.000000  
13      16666.666667  
14      12750.000000  
15      13333.333333  
16      18666.666667  
17      16923.076923

In [113]:
final_df[final_df['ori_des_pair'] == 'AA-BB']

ori_des_pair air_type  NO. of trips will take  cost_per_trip  \
0        AA-BB     A320                       3        18750.0   
1        AA-BB     A330                       2        25000.0   
2        AA-BB     B737                       3        19125.0   
3        AA-BB     B747                       2        40000.0   

   Total_cost(in $)  
0           56250.0  
1           50000.0  
2           57375.0  
3           80000.0

**From above dataframe we can see that for city pair of 'AA-BB'  'A330' Aircraft Type is best suited**

In [114]:
final_df[final_df['ori_des_pair'] == 'BB-CC']

ori_des_pair air_type  NO. of trips will take  cost_per_trip  \
4        BB-CC     A330                       2   54166.666667   
5        BB-CC     B747                       2   86666.666667   

   Total_cost(in $)  
4     108333.333333  
5     173333.333333

**From above dataframe we can see that for city pair of 'BB-CC'  'A330' Aircraft Type is best suited**

In [115]:
final_df[final_df['ori_des_pair'] == 'CC-AA']

ori_des_pair air_type  NO. of trips will take  cost_per_trip  \
6         CC-AA     A320                       2    2500.000000   
7         CC-AA     A330                       2    3333.333333   
8         CC-AA     B737                       2    2550.000000   
9         CC-AA     B747                       1    5333.333333   
10        CC-AA     Q400                       4    1866.666667   
11        CC-AA    ATR72                       4    1692.307692   

    Total_cost(in $)  
6        5000.000000  
7        6666.666667  
8        5100.000000  
9        5333.333333  
10       7466.666667  
11       6769.230769

**From above dataframe we can see that for city pair of 'CC-AA'  'A320' Aircraft Type is best suited**

In [116]:
final_df[final_df['ori_des_pair'] == 'AA-DD']

ori_des_pair air_type  NO. of trips will take  cost_per_trip  \
12        AA-DD     A320                       2    6250.000000   
13        AA-DD     A330                       2    8333.333333   
14        AA-DD     B737                       2    6375.000000   
15        AA-DD     B747                       1   13333.333333   
16        AA-DD     Q400                       4    4666.666667   
17        AA-DD    ATR72                       4    4230.769231   

    Total_cost(in $)  
12      12500.000000  
13      16666.666667  
14      12750.000000  
15      13333.333333  
16      18666.666667  
17      16923.076923

**From above dataframe we can see that for city pair of 'AA-DD'  'A320' Aircraft Type is best suited**